shared/gateway/api_middleware.go

```diff
@@ -0,0 +1,463 @@
+package gateway
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+	butil "github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/grpcutils"
+	"github.com/wealdtech/go-bytesutil"
+)
+
+// ApiProxyMiddleware is a proxy between an Eth2 API HTTP client and gRPC-gateway.
+// The purpose of the proxy is to handle HTTP requests and gRPC responses in such a way that:
+//   - Eth2 API requests can be handled by gRPC-gateway correctly
+//   - gRPC responses can be returned as spec-compliant Eth2 API responses
+type ApiProxyMiddleware struct {
+	GatewayAddress string
+	ProxyAddress   string
+	router         *mux.Router
+}
+
+type endpointData struct {
+	postRequest interface{}
+	getResponse interface{}
+	err         ErrorJson
+}
+
+// fieldProcessor applies the processing function f to a value when the tag is present on the field.
+type fieldProcessor struct {
+	tag string
+	f   func(value reflect.Value) error
+}
+
+// Run starts the proxy, registering all proxy endpoints on ApiProxyMiddleware.ProxyAddress.
+func (m *ApiProxyMiddleware) Run() error {
+	m.router = mux.NewRouter()
+
+	m.handleApiEndpoint("/eth/v1/beacon/genesis")
```

Comment:
this list of endpoints should be passed into this function. The shared/ package is meant to be generic and does not need to know anything about the world outside it

Reply:
Endpoints extracted to `/beacon-chain/rpc`.

I don't think there is any way to enforce endpoint registration. We need documentation for this. I am thinking about writing a dev wiki page.